src/java/org/apache/cassandra/index/sai/disk/v5/V5VectorPostingsWriter.java

```diff
@@ -419,32 +434,110 @@ public static <T> RemappedPostings remapPostings(Map<VectorFloat<?>, ? extends V
 
         // create the mapping
         if (structure == Structure.ZERO_OR_ONE_TO_MANY)
-            return createGenericMapping(ordinalMap.keySet(), maxOldOrdinal, maxRow);
-        return new RemappedPostings(structure, maxNewOrdinal, maxRow, ordinalMap, extraPostings);
+            return createGenericRenumberedMapping(ordinalMap.keySet(), maxOldOrdinal, maxRow);
+        var ordinalMapper = new BiMapMapper(maxNewOrdinal, ordinalMap);
+        return new RemappedPostings(structure, maxNewOrdinal, maxRow, ordinalMap, extraPostings, ordinalMapper);
     }
 
     /**
      * return an exhaustive zero-to-many mapping with the live ordinals renumbered sequentially
      */
-    private static RemappedPostings createGenericMapping(Set<Integer> liveOrdinals, int maxOldOrdinal, int maxRow)
+    private static RemappedPostings createGenericRenumberedMapping(Set<Integer> liveOrdinals, int maxOldOrdinal, int maxRow)
     {
-        var sequentialMap = new Int2IntHashMap(maxOldOrdinal, 0.65f, Integer.MIN_VALUE);
+        var oldToNew = new Int2IntHashMap(maxOldOrdinal, 0.65f, Integer.MIN_VALUE);
         int nextOrdinal = 0;
         for (int i = 0; i <= maxOldOrdinal; i++) {
             if (liveOrdinals.contains(i))
-                sequentialMap.put(i, nextOrdinal++);
+                oldToNew.put(i, nextOrdinal++);
         }
-        return new RemappedPostings(nextOrdinal - 1, maxRow, sequentialMap);
+        return new RemappedPostings(Structure.ZERO_OR_ONE_TO_MANY,
+                                    nextOrdinal - 1,
+                                    maxRow,
+                                    null,
+                                    null,
+                                    new OrdinalMapper.MapMapper(oldToNew));
     }
 
     /**
-     * return an exhaustive zero-to-many mapping for v2 postings, which never contain missing ordinals
-     * since deleted vectors are only removed from the index in its next compaction
+     * return an exhaustive zero-to-many mapping with no renumbering
      */
-    public static <T> RemappedPostings createGenericV2Mapping(Map<VectorFloat<?>, ? extends VectorPostings<T>> postingsMap)
+    public static <T> RemappedPostings createGenericIdentityMapping(Map<VectorFloat<?>, ? extends VectorPostings<T>> postingsMap)
     {
-        int maxOldOrdinal = postingsMap.size() - 1;
+        var maxOldOrdinal = postingsMap.values().stream().mapToInt(VectorPostings::getOrdinal).max().orElseThrow();
         int maxRow = postingsMap.values().stream().flatMap(p -> p.getRowIds().stream()).mapToInt(i -> i).max().orElseThrow();
-        return createGenericMapping(IntStream.range(0, postingsMap.size()).boxed().collect(Collectors.toSet()), maxOldOrdinal, maxRow);
+        var presentOrdinals = new FixedBitSet(maxOldOrdinal + 1);
+        for (var entry : postingsMap.entrySet())
+            presentOrdinals.set(entry.getValue().getOrdinal());
```

Comment:
Is it worth collapsing these two iterations into a single loop since postingsMap is on disk?

Reply:
I was trying to avoid spending time on optimizing the Generic case, but we just had that colbert issue today that shows that it does happen, so you're probably right

Reply:
actually in this case we need maxOldOrdinal to determine the size of the BitSet, so we'd have to switch to a less efficient representation to do it in a single pass.  i'd prefer to leave it alone.